File: experiments/2d_visualization/cpp/test_aisgng_roman_triple_ring.cpp
/**
 * @file test_aisgng_roman_triple_ring.cpp
 * @brief Test AiS-GNG RO-MAN 2023 (single threshold) on triple ring data
 */

#include <chrono>
#include <cmath>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <random>
#include <sstream>
#include <string>
#include <vector>

#include <Eigen/Core>
#include "ais_gng_roman.hpp"

namespace fs = std::filesystem;

std::vector<Eigen::Vector2f> sample_triple_ring(int n_samples, unsigned int seed) {
    std::mt19937 rng(seed);
    std::uniform_real_distribution<float> angle_dist(0.0f, 2.0f * M_PI);

    std::vector<std::tuple<float, float, float, float>> rings = {
        {0.50f, 0.23f, 0.06f, 0.14f},
        {0.27f, 0.68f, 0.06f, 0.14f},
        {0.73f, 0.68f, 0.06f, 0.14f},
    };

    std::vector<Eigen::Vector2f> samples;
    samples.reserve(n_samples);
    int samples_per_ring = n_samples / 3;

    for (const auto& [cx, cy, r_inner, r_outer] : rings) {
        std::uniform_real_distribution<float> radius_dist(r_inner, r_outer);
        for (int i = 0; i < samples_per_ring; ++i) {
            float angle = angle_dist(rng);
            float radius = radius_dist(rng);
            float x = cx + radius * std::cos(angle);
            float y = cy + radius * std::sin(angle);
            samples.emplace_back(x, y);
        }
    }
    return samples;
}

void save_samples(const std::vector<Eigen::Vector2f>& samples, const std::string& path) {
    std::ofstream file(path);
    file << "x,y\n";
    file << std::fixed << std::setprecision(6);
    for (const auto& p : samples) {
        file << p.x() << "," << p.y() << "\n";
    }
}

void save_graph(const ais_gng_roman::AiSGNGRoman2f& model,
                const std::string& nodes_path, const std::string& edges_path) {
    std::vector<Eigen::Vector2f> nodes;
    std::vector<std::pair<int, int>> edges;
    model.get_graph(nodes, edges);

    {
        std::ofstream file(nodes_path);
        file << "x,y\n";
        file << std::fixed << std::setprecision(6);
        for (const auto& n : nodes) {
            file << n.x() << "," << n.y() << "\n";
        }
    }
    {
        std::ofstream file(edges_path);
        file << "i,j\n";
        for (const auto& e : edges) {
            file << e.first << "," << e.second << "\n";
        }
    }
}

int main(int argc, char* argv[]) {
    std::string output_dir = "aisgng_roman_triple_ring_output";
    int n_samples = 1500;
    int n_iterations = 5000;
    unsigned int seed = 42;
    int frame_interval = 50;

    if (argc > 1) output_dir = argv[1];
    if (argc > 2) n_samples = std::stoi(argv[2]);
    if (argc > 3) n_iterations = std::stoi(argv[3]);
    if (argc > 4) seed = static_cast<unsigned int>(std::stoi(argv[4]));

    fs::create_directories(output_dir);

    std::cout << "AiS-GNG RO-MAN 2023 Triple Ring Test (C++)\n";
    std::cout << "==========================================\n";
    std::cout << "Output directory: " << output_dir << "\n";
    std::cout << "Samples: " << n_samples << "\n";
    std::cout << "Iterations: " << n_iterations << "\n\n";

    auto samples = sample_triple_ring(n_samples, seed);
    save_samples(samples, output_dir + "/samples.csv");

    ais_gng_roman::AiSGNGRomanParams params;
    params.max_nodes = 100;
    params.lambda = 100;
    params.eps_b = 0.08f;
    params.eps_n = 0.008f;
    params.max_age = 88;
    params.utility_k = 1000.0f;
    params.kappa = 10;
    params.theta_ais = 0.10f;  // Single threshold (RO-MAN 2023)

    std::cout << "RO-MAN 2023 Parameters:\n";
    std::cout << "  theta_ais: " << params.theta_ais << " (single threshold)\n\n";

    ais_gng_roman::AiSGNGRoman2f model(params, seed);
    model.init();

    std::cout << "Training...\n";
    std::vector<int> frame_iterations;

    auto start_time = std::chrono::high_resolution_clock::now();

    auto callback = [&](const ais_gng_roman::AiSGNGRoman2f& m, int iter) {
        if (iter % frame_interval == 0 || iter == n_iterations - 1) {
            std::stringstream ss;
            ss << std::setfill('0') << std::setw(5) << iter;
            save_graph(m, output_dir + "/nodes_" + ss.str() + ".csv",
                       output_dir + "/edges_" + ss.str() + ".csv");
            frame_iterations.push_back(iter);

            std::cout << "Iteration " << iter << ": "
                      << m.num_nodes() << " nodes, +" << m.n_ais_additions << " AiS\n";
        }
    };

    model.train(samples, n_iterations, callback);

    auto end_time = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);

    save_graph(model, output_dir + "/final_nodes.csv", output_dir + "/final_edges.csv");

    {
        std::ofstream file(output_dir + "/frames.csv");
        file << "iteration\n";
        for (int iter : frame_iterations) file << iter << "\n";
    }

    {
        std::ofstream file(output_dir + "/metadata.csv");
        file << "key,value\n";
        file << "algorithm,AiS-GNG-RO-MAN\n";
        file << "n_iterations," << n_iterations << "\n";
        file << "time_ms," << duration.count() << "\n";
        file << "final_nodes," << model.num_nodes() << "\n";
        file << "final_edges," << model.num_edges() << "\n";
        file << "n_ais_additions," << model.n_ais_additions << "\n";
    }

    std::cout << "\nComplete! Time: " << duration.count() << " ms\n";
    std::cout << "Final: " << model.num_nodes() << " nodes, AiS: " << model.n_ais_additions << "\n";

    return 0;
}

File: experiments/2d_visualization/test_aisgng_roman_triple_ring.py
"""Test AiS-GNG RO-MAN 2023 version on triple ring data.

RO-MAN 2023 version uses a single threshold θ_AiS instead of range [θ_min, θ_max].
Condition: ||v_t - h_s1|| < θ_AiS AND ||v_t - h_s2|| < θ_AiS

Usage:
    python test_aisgng_roman_triple_ring.py
"""

import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
from PIL import Image

sys.path.insert(0, str(Path(__file__).parents[2] / "algorithms" / "ais_gng" / "python"))
sys.path.insert(0, str(Path(__file__).parents[2] / "data" / "2d"))

from model_roman import AiSGNGRoman, AiSGNGRomanParams
from sampler import sample_triple_ring

TRIPLE_RING_PARAMS = [
    (0.50, 0.23, 0.06, 0.14),
    (0.27, 0.68, 0.06, 0.14),
    (0.73, 0.68, 0.06, 0.14),
]


def draw_triple_ring_background(ax) -> None:
    """Draw triple ring background."""
    theta = np.linspace(0, 2 * np.pi, 100)
    for cx, cy, r_inner, r_outer in TRIPLE_RING_PARAMS:
        outer_x = cx + r_outer * np.cos(theta)
        outer_y = cy + r_outer * np.sin(theta)
        inner_x = cx + r_inner * np.cos(theta)
        inner_y = cy + r_inner * np.sin(theta)
        ax.fill(outer_x, outer_y, color="lightblue", alpha=0.3)
        ax.fill(inner_x, inner_y, color="white")


def create_frame(
    ax,
    points: np.ndarray,
    nodes: np.ndarray,
    edges: list[tuple[int, int]],
    iteration: int,
    n_ais_additions: int = 0,
    n_utility_removals: int = 0,
) -> None:
    """Create a single frame for visualization."""
    ax.clear()
    draw_triple_ring_background(ax)

    ax.scatter(points[:, 0], points[:, 1], c="skyblue", s=3, alpha=0.3)

    for i, j in edges:
        ax.plot(
            [nodes[i, 0], nodes[j, 0]],
            [nodes[i, 1], nodes[j, 1]],
            "r-",
            linewidth=1.5,
            alpha=0.7,
        )

    ax.scatter(nodes[:, 0], nodes[:, 1], c="red", s=50, zorder=5)

    ax.set_xlim(0, 1)
    ax.set_ylim(1, 0)
    ax.set_aspect("equal")
    ax.set_title(
        f"AiS-GNG RO-MAN - Iter {iteration} ({len(nodes)} nodes, "
        f"+{n_ais_additions} AiS, -{n_utility_removals} utility)"
    )


def run_experiment(
    n_samples: int = 1500,
    n_iterations: int = 5000,
    gif_frames: int = 100,
    output_gif: str = "aisgng_roman_triple_ring_growth.gif",
    output_final: str = "aisgng_roman_triple_ring_final.png",
    seed: int = 42,
    theta_ais: float = 0.10,
) -> None:
    """Run AiS-GNG RO-MAN experiment."""
    np.random.seed(seed)

    print(f"Sampling {n_samples} points from triple ring...")
    points = sample_triple_ring(n_samples=n_samples, seed=seed)

    params = AiSGNGRomanParams(
        max_nodes=100,
        lambda_=100,
        eps_b=0.08,
        eps_n=0.008,
        alpha=0.5,
        beta=0.005,
        chi=0.005,
        max_age=100,
        utility_k=1000.0,
        kappa=10,
        theta_ais=theta_ais,  # Single threshold (RO-MAN 2023)
    )
    gng = AiSGNGRoman(n_dim=2, params=params, seed=seed)

    frames = []
    frame_interval = max(1, n_iterations // gif_frames)

    fig, ax = plt.subplots(figsize=(8, 8), facecolor="white")

    def callback(model, iteration):
        if iteration % frame_interval == 0 or iteration == n_iterations - 1:
            nodes, edges = model.get_graph()
            create_frame(
                ax, points, nodes, edges, iteration,
                model.n_ais_additions, model.n_utility_removals
            )
            fig.canvas.draw()

            img = Image.frombuffer(
                "RGBA", fig.canvas.get_width_height(), fig.canvas.buffer_rgba()
            )
            frames.append(img.convert("RGB"))
            print(
                f"Iteration {iteration}: {len(nodes)} nodes, {len(edges)} edges, "
                f"+{model.n_ais_additions} AiS"
            )

    print(f"Training AiS-GNG RO-MAN for {n_iterations} iterations...")
    print(f"Single threshold: theta_ais={params.theta_ais}")
    gng.train(points, n_iterations=n_iterations, callback=callback)

    nodes, edges = gng.get_graph()
    create_frame(
        ax, points, nodes, edges, n_iterations,
        gng.n_ais_additions, gng.n_utility_removals
    )
    plt.savefig(output_final, dpi=150, bbox_inches="tight", facecolor="white")
    print(f"Saved final result: {output_final}")

    if frames:
        frames.extend([frames[-1]] * 10)
        frames[0].save(
            output_gif,
            save_all=True,
            append_images=frames[1:],
            duration=100,
            loop=0,
        )
        print(f"Saved GIF: {output_gif}")

    plt.close(fig)
    print(f"Done! AiS additions: {gng.n_ais_additions}")


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Test AiS-GNG RO-MAN on triple ring")
    parser.add_argument("-n", "--n-samples", type=int, default=1500)
    parser.add_argument("--iterations", type=int, default=5000)
    parser.add_argument("--frames", type=int, default=100)
    parser.add_argument("--output-gif", type=str, default="aisgng_roman_triple_ring_growth.gif")
    parser.add_argument("--output-final", type=str, default="aisgng_roman_triple_ring_final.png")
    parser.add_argument("--seed", type=int, default=42)
    parser.add_argument("--theta-ais", type=float, default=0.10, help="Single AiS threshold")

    args = parser.parse_args()

    run_experiment(
        n_samples=args.n_samples,
        n_iterations=args.iterations,
        gif_frames=args.frames,
        output_gif=args.output_gif,
        output_final=args.output_final,
        seed=args.seed,
        theta_ais=args.theta_ais,
    )
